File: Rust/15. 3Sum.rs
use std::collections::HashSet;

impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut set: HashSet<Vec<i32>> = HashSet::new();
        let mut left: usize;
        let mut right: usize;
        nums.sort();

        for i in 0..nums.len() {
            left = i + 1;
            right = nums.len() - 1;
            while left < right {
                match nums[i] + nums[left] + nums[right] {
                    sum if sum == 0 => {
                        set.insert(vec![nums[i], nums[left], nums[right]]);
                        right -= 1;
                        left += 1;
                    }
                    sum if sum > 0 => {
                        right -= 1;
                    }
                    _ => {
                        left += 1;
                    }
                }
            }
        }

        set.into_iter().collect()
    }
}

impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut left: usize;
        let mut right: usize;
        nums.sort();

        for i in 0..nums.len() - 2 {
            // if first number is already larger then target then target no longer exists
            if nums[i] > 0 {
                break;
            }

            // skip the repeats
            if i > 0 && nums[i] == nums[i - 1] {
                continue;
            }

            left = i + 1;
            right = nums.len() - 1;

            while left < nums.len() && left < right {
                match nums[i] + nums[left] + nums[right] {
                    sum if sum == 0 => {
                        result.push(vec![nums[i], nums[left], nums[right]]);

                        while left < nums.len() - 1 && nums[left] == nums[left + 1] {
                            left += 1;
                        }

                        while right > 0 && nums[right] == nums[right - 1] {
                            right -= 1;
                        }

                        right -= 1;
                        left += 1;
                    }
                    sum if sum > 0 => {
                        right -= 1;
                    }
                    _ => {
                        left += 1;
                    }
                }
            }
        }
        result
    }
}

File: Java/15. 3Sum.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        int n = nums.length;
        int l, r, sum;
        for (int i = 0; i < n - 2; i++) {

            if (i == 0 || nums[i - 1] < nums[i]) {

                l = i + 1;
                r = n - 1;

                while (l < r) {
                    sum = nums[i] + nums[l] + nums[r];
                    if (sum == 0) {
                        result.add(List.of(nums[i], nums[l], nums[r]));

                        while (l < r && nums[l] == nums[l + 1])
                            l++;

                        while (l < r && nums[r] == nums[r - 1])
                            r--;

                        l++;
                        r--;

                    } else if (sum < 0)
                        l++;
                    else
                        r--;
                }
            }
        }

        return result;
    }
}